src/expr/src/interpret.rs

```diff
@@ -346,43 +347,50 @@ impl<'a> ResultSpec<'a> {
         self.fallible
     }
 
-    /// This function "maps" a function across the set of valid datums, returning the set of possible
-    /// results.
+    /// This method "maps" a function across the `ResultSpec`.
     ///
-    /// If we actually stored a set of every possible value, this could by done by passing
-    /// each datum to the function one-by-one and unioning the resulting sets. This is possible
-    /// when our values set is empty or     contains a single datum, but when it contains a range,
-    /// we need a little metadata about the function to be able to infer a useful result.
+    /// As mentioned above, `ResultSpec` represents an approximate set of results.
+    /// If we actually stored each result in the set, `flat_map` could be implemented by passing
+    /// each result to the function one-by-one and unioning the resulting sets. This is possible
+    /// when our values set is empty or contains a single datum, but when it contains a range,
+    /// we can't enumerate all possible values of the set. We handle this by:
+    /// - tracking whether the function is monotonic, in which case we can map the range by just
+    ///   mapping the endpoints;
+    /// - using a safe default when we can't infer a tighter bound on the set, eg. [Self::anything].
     fn flat_map(
-        self,
+        &self,
         is_monotonic: Monotonic,
-        mut datum_map: impl FnMut(Datum<'a>) -> ResultSpec<'a>,
+        mut result_map: impl FnMut(Result<Datum<'a>, EvalError>) -> ResultSpec<'a>,
     ) -> ResultSpec<'a> {
         let null_spec = if self.nullable {
-            datum_map(Datum::Null)
+            result_map(Ok(Datum::Null))
         } else {
             ResultSpec::nothing()
         };
 
-        let error_spec = ResultSpec {
-            fallible: self.fallible,
-            ..Self::nothing()
+        let error_spec = if self.fallible {
+            // Since we only care about whether / not an error is possible, and not the specific
+            // error, create an arbitrary error here.
+            // NOTE! This assumes that functions do not discriminate on the type of the error.
```

Comment:
I suggest to replicate this not somewhere in the `eval` code so people can stick to this assumption when they add new / evolve existing `eval` functions.